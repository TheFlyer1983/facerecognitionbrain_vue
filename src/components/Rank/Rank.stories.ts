import { ArgTypes, Story } from '@storybook/vue3';
import Rank from './Rank.vue';

export default {
  title: 'Rank',
  component: Rank
};

const Template: Story = (args) => ({
  components: { Rank },
  setup() {
    return { args };
  },
  template: `<Rank :user="args.user" :userRank="args.userRank"/>`
});

export const Default = Template.bind({});

Default.args = {
  user: {
    id: '1',
    name: 'Paul',
    email: 'test@test.com',
    entries: 1,
    joined: '2021-04-09T00:00:00.000Z',
    pet: 'Ruby',
    age: 37
  },
  userRank: 'ğŸ˜€'
};
Default.argTypes = {
  user: {
    description: 'The current logged in user'
  },
  userRank: {
    description: 'The rank of the user'
  }
} as ArgTypes;
