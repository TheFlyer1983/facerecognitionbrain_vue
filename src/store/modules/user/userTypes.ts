import { ActionContext } from 'vuex';

export type UserActionContext = ActionContext<UserState, {}>

export interface UserState {
  isSignedIn: boolean;
  isProfileOpen: false;
  user: User | null;
  token: string;
  rank: string;
}

export interface User {
  id: string;
  name: string;
  email: string;
  entries: number;
  joined: string;
  pet: string;
  age: string;
}
