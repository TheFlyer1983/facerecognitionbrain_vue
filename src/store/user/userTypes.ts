export interface UserState {
  isSignedIn: boolean;
  isProfileOpen: boolean;
  id: string;
  user: User | null;
  token: string;
  rank: string;
}

export interface User {
  name: string;
  email: string;
  entries: number;
  joined: string | number | Date;
  pet?: string;
  age?: number;
  role?: string;
}

export interface UpdateInfo {
  name?: string;
  pet?: string;
  age?: number;
}
