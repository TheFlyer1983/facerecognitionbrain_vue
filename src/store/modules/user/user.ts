import { UserState, UserActionContext } from './userTypes';

import { endpoints } from '@/constants';
import { LoginInfo, RegisterInfo } from '@/types';
import {
  getAuthTokenInSession,
  saveAuthTokenInSession
} from '@/functions/storageFunctions';

export const state: UserState = {
  isSignedIn: false,
  isProfileOpen: false,
  user: null,
  token: ''
};

export const getters = {
  getUser(state: UserState): UserState['user'] {
    return state.user;
  },
  getToken(state: UserState): UserState['token'] {
    return state.token;
  },
  getIsSignedIn(state: UserState): UserState['isSignedIn'] {
    return state.isSignedIn;
  }
};

export const mutations = {
  setUserId(state: UserState, payload: string): void {
    state.user = { ...state.user, id: payload } as UserState['user'];
  },

  setUser(state: UserState, payload: UserState['user']): void {
    state.user = payload;
  },

  toggleSignIn(state: UserState, payload: UserState['isSignedIn']): void {
    state.isSignedIn = payload;
  },

  setToken(state: UserState, payload: UserState['token']): void {
    state.token = payload;
  }
};

export const actions = {
  async login(
    { commit, dispatch }: UserActionContext,
    payload: LoginInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.signin, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.token as string);
      commit('setToken', result.token);
      commit('setUserId', result.userId);

      const success = await dispatch('getUser', result.userId as string);
      if (!success) throw new Error('Error');

      return true;
    } catch (error) {
      console.log(error, 'In the error block');
      return false;
    }
  },
  async getUser({ commit }: UserActionContext, payload: string) {
    const requestURL = endpoints.getProfile.replace(':id', payload);
    const token = getAuthTokenInSession();
    try {
      if (!token) throw new Error('No valid token');
      const response = await fetch(requestURL, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          Authorization: token
        }
      });
      const result = await response.json();
      if (!response.ok) throw new Error(result);

      commit('setUser', result);
      commit('toggleSignIn', true);
      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  async registerUser(
    { commit }: UserActionContext,
    payload: RegisterInfo
  ): Promise<boolean> {
    try {
      const response = await fetch(endpoints.register, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });

      const result = await response.json();

      if (!response.ok) throw new Error(result as string);

      saveAuthTokenInSession(result.session.token as string);
      commit('setUser', result.register.user);

      return true;
    } catch (error) {
      console.log(error);
      return false;
    }
  },
  getToken({ commit, dispatch }: UserActionContext) {
    const token = getAuthTokenInSession();

    if (token) {
      commit('setToken', token);
      dispatch('authenticated');
    }
  },
  async authenticated({ state, dispatch }: UserActionContext) {
    const token = state.token;

    if (token) {
      try {
        const response = await fetch(endpoints.signin, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: token
          }
        });

        const result = await response.json();

        if (!response.ok) throw new Error(result as string);

        dispatch('getUser', result.id);
      } catch (error) {
        console.log(error);
      }
    }
  }
};

const userModule = {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
};

export default userModule;
