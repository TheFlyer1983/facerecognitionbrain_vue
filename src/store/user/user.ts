import { defineStore } from 'pinia';
import { UpdateInfo, User, UserState } from './userTypes';
import {
  LoginInfo,
  LoginResponse,
  RegisterInfo,
  RegisterResponse
} from '@/types';
import request from '@/functions/request';
import { endpoints } from '@/constants';
import {
  getAuthTokenInSession,
  removeAuthTokenFromSession,
  saveAuthTokenInSession
} from '@/functions/storageFunctions';

export const useUserStore = defineStore('user', {
  state: (): UserState => ({
    isSignedIn: false,
    isProfileOpen: false,
    id: '',
    user: null,
    token: '',
    rank: ''
  }),

  actions: {
    async login(payload: LoginInfo) {
      payload = { ...payload, returnSecureToken: true };

      try {
        const response = await request.post<LoginResponse>(
          endpoints.signin,
          payload,
          {
            params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY }
          }
        );

        saveAuthTokenInSession(response.data.idToken);
        this.token = response.data.idToken;
        this.id = response.data.localId;

        const success = await this.getUser(response.data.localId);
        if (!success) throw new Error('Error');

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async getUser(userId: string) {
      const requestURL = endpoints.profile.replace(':id', userId);

      try {
        const response = await request.get<User>(requestURL, {
          params: { auth: this.token }
        });

        this.user = response.data;
        this.isSignedIn = true;

        //this.getRank();

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async registerUser(payload: RegisterInfo) {
      const registerPayload = {
        email: payload.email,
        password: payload.password,
        returnSecureToken: true
      };
      try {
        const response = await request.post<RegisterResponse>(
          endpoints.register,
          registerPayload,
          {
            params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY }
          }
        );

        saveAuthTokenInSession(response.data.idToken);
        this.token = response.data.idToken;
        this.id = response.data.localId;
        await this.createProfile(payload.name);

        return true;
      } catch (error) {
        console.error(error);
        return false;
      }
    },

    async createProfile(name: string) {
      const profilePayload = {
        name,
        entries: 0,
        joined: new Date()
      };

      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        const response = await request.put(requestURL, profilePayload, {
          params: { auth: this.token }
        });

        this.user = response.data;
      } catch (error) {
        console.log(error);
      }
    },

    getToken() {
      const token = getAuthTokenInSession();

      if (token) {
        this.token = token;

        this.authenticated();
      }
    },

    async authenticated() {
      if (this.token) {
        try {
          const response = await request.post(endpoints.signin);

          this.getUser(response.data.id);
        } catch (error) {
          console.error(error);
        }
      }
    },

    async getRank() {
      const entries = this.user?.entries;

      try {
        const response = await request.post(endpoints.rank, { entries });
        this.rank = response.data;
      } catch (error) {
        console.error(error);
      }
    },

    signout() {
      removeAuthTokenFromSession();
      this.$reset();
    },

    async updateUser(payload: UpdateInfo) {
      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        await request.put(requestURL, payload);

        this.getUser(this.id as string);
      } catch (error) {
        console.error(error);
      }
    },

    async deleteUser() {
      const requestURL = endpoints.profile.replace(':id', this.id);
      try {
        await request.delete(requestURL, { params: { auth: this.token } });

        await request.post(
          endpoints.delete,
          { idToken: this.id },
          { params: { key: import.meta.env.VITE_APP_FIREBASE_API_KEY } }
        );

        this.signout();
      } catch (error) {
        console.error(error);
      }
    }
  }
});
