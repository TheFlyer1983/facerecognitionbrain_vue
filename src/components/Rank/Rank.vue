<script setup lang="ts">
import { User, UserState } from '@/store/user/userTypes';
import { NonNullableObject } from '@/types';

defineProps<{
  user: NonNullableObject<User>;
  userRank: UserState['rank'];
}>();
</script>

<template>
  <div class="rank d-flex flex-column align-items-center">
    <div>{{ user.name }}, your current entry count is...</div>
    <div class="rank-ranking">{{ user.entries }}</div>
    <div>Rank Badge: {{ userRank }}</div>
  </div>
</template>

<style scoped lang="scss">
.rank {
  text-align: center;
  color: white;
  font-size: 1.5rem;

  &-ranking {
    font-size: 3rem;
  }
}
</style>
