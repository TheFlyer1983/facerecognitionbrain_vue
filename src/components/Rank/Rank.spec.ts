import { shallowMount, VueWrapper } from '@vue/test-utils';
import { beforeEach, describe, it, expect, vi, beforeAll } from 'vitest';
import { reactive } from 'vue';
import { useStore } from 'vuex';

import Rank from './Rank.vue';
import { UserState } from '@/store/modules/user/userTypes';

const mockProps = {
  user: {
    id: 1,
    name: 'Paul',
    email: 'test@test.com',
    entries: 1,
    joined: '2021-04-09T00:00:00.000Z',
    pet: 'Ruby',
    age: 37
  }
};

vi.mock('vuex');

const mockedUseStore = vi.mocked<() => Partial<typeof useStore>>(useStore);

describe('Given the `Rank` component', () => {
  const render = () => shallowMount(Rank, { props: { user: mockProps.user } });
  let wrapper: VueWrapper<InstanceType<typeof Rank>>;

  const mockedGetters = {
    ['user/getRank']: 'ðŸ˜ƒ' as UserState['rank']
  };

  const storeMock = reactive({
    getters: mockedGetters
  });

  beforeAll(() => {
    mockedUseStore.mockReturnValue(storeMock);
  });

  describe('when it is rendered', () => {
    beforeEach(() => {
      wrapper = render();
    });

    it('should match the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
