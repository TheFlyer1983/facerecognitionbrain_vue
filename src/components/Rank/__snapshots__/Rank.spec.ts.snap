// Vitest Snapshot v1

exports[`Given the \`Rank\` component > when it is rendered > should match the snapshot 1`] = `
<div
  class="rank d-flex flex-column align-items-center"
  data-v-e828e83c=""
>
  <div
    data-v-e828e83c=""
  >
    Paul, your current entry count is...
  </div>
  <div
    class="rank-ranking"
    data-v-e828e83c=""
  >
    1
  </div>
  <div
    data-v-e828e83c=""
  >
    Rank Badge: ğŸ˜€
  </div>
</div>
`;
